### DISCUSSION TITLE
Open-source the new MSBuild targets/tasks?

### DISCUSSION THREAD

**Discussion by wjk** [microsoft/WindowsAppSDK#899] [Q&A]
The Reunion 0.8 NuGet packages contain a sizable amount of brand-new, never-before-seen MSBuild code, primarily in the area of MSIX packaging. Would I be able to ask you for a guesstimate of when this code (both the targets and the source code for the tasks) will be posted to GitHub? (My interest in this code is not purely academic. I currently have to add an MSBuild monkey-patch to workaround a limitation in the PRI generation targets, and I would like to be able to submit these changes upstream in a PR.)

Note that I am _not_ including `Microsoft.UI.Xaml.Markup.Compiler.interop.targets` and its contents in this request, since I was told in https://github.com/microsoft/microsoft-ui-xaml/issues/1433#issuecomment-851035061 exactly how messy that particular codebase is, and I do not want to have this request be blocked on that one. Thanks!

**Comment by Scottj1s** [comment]
Hi William, most of the interesting MSBuild code is maintained in the closed source WinUI 3 repo. There has been discussion about open-sourcing WinUI 3 (as we've done for WinUI), but nothing concrete as yet.  That said, if you have a fix, feel free to attach it to this issue and I'll see that it propagates back to the source.

**Reply by wjk** [reply]
> There has been discussion about open-sourcing WinUI 3 (as we've done for WinUI), but nothing concrete as yet.

I’m sorry, but I was told that from the beginning the plan was always to open-source WinUI 3. Most of the discussion in the WinUI 3 repo is made under the assumption that we will see sources eventually. [The WinUI 3 roadmap corroborates this](https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md#benefits-of-winui). To hear that the sources will _not_ be released in the foreseeable future is very bad news.

**Reply by andrewleader** [reply]
@wjk we do indeed have plans to open source as much of Project Reunion components as possible. We'll be sharing our open source roadmap plans in a month or two. However, for the 1.0 timeframe, we are focusing on some other key priorities like full unpackaged app support and more. We cannot do everything in 1.0 and are focusing on top customer requests, but I'll add your name to the list of requests for open sourcing (which is certainly a popular request that we hope to make more progress on soon, some components like dynamic dependencies and AppLifecycle are already fully open source).

**Reply by wjk** [reply]
@andrewleader Thanks for getting back to me. I fully support making WinUI 100% functional before making it open-source. Just as long as the open-sourcing hasn’t been cancelled entirely, I’m good. 😄

------

@Scottj1s My request regarding the MSBuild targets is as follows:

Whenever I use a PRI file, I want to embed all binary resources (usually image files) into the PRI instead of having them loose on the filesystem. By default, `@(PRIResource)` items are added by path reference. I must copy the items into the output directory or the resource loader would throw. While XBF files are automatically embedded (and amen to that!), embedding other files requires that I use the following code:

```xml
<Target Name="UpdateEmbedPriFiles">
    <ItemGroup>
      <Temp Include="@(PRIResource)" Condition="'%(PRIResource.Embed)' == 'true'" />
      <PRIResource Remove="@(Temp)" />

      <_PriOutputsUnexpanded Include="@(Temp)">
          <OutputGroup>EmbedOutputGroupForPackaging</OutputGroup>
          <ProjectName>$(MSBuildProjectName)</ProjectName>
      </_PriOutputsUnexpanded>

      <Temp Remove="@(Temp)" />
    </ItemGroup>
</Target>

<PropertyGroup>
  <GetPriOutputsDependsOn>$(GetPriOutputsDependsOn);UpdateEmbedPriFiles</GetPriOutputsDependsOn>
</PropertyGroup>
```

This target adds `PRIResource` items marked with the `Embed` metadata item (which I made up for purposes of this code) to the appropriate private item group to have them embedded when the PRI file is generated. While the above code works, I would prefer to have a way to do this without needing to rely on implementation details. Thanks!

**Reply by huichen123** [reply]
To your specific PRI problem, if you use Reunion/MRTCore for PRI file, [Microsoft.ApplicationModel.Resources.PriGen.targets]( https://github.com/microsoft/ProjectReunion/blob/main/dev/MRTCore/packaging/Microsoft.ApplicationModel.Resources.PriGen.targets)  is already open source.

**Reply by wjk** [reply]
@huichen123 Unfortunately, the open-source targets file are not compatible with the PRI-file generation targets used by WinUI. If I enable both, MSBuild emits the below error (line-wrapped for readability). Therefore, since I cannot use the open-source targets with my project, I cannot submit a PR. (This is not to mention the fact that the PRI-file targets used by WinUI are based on a completely different design than the open-source targets. I find the WinUI version to be far simpler and easier to understand.)

```
D:\Projects\AeroOffice\packages\microsoft.projectreunion.foundation\0.8.0-preview\
build\Microsoft.ApplicationModel.Resources.PriGen.targets(740,5):
error MSB4044: The "CreatePriConfigXmlForFullIndex" task was not
given a value for the required parameter "DefaultResourceLanguage".

**Reply by huichen123** [reply]
@wjk I assume both MRT and WinUI targets rooted from AppXPackage.targets. Please file a new issue if you think WinUI version is simpler and easier to understand. As for the error you see, can you set `<DefaultLanguage>en-US</DefaultLanguage>` in the project file?
